tests/test_zarr_compat.py

```diff
@@ -5,26 +5,104 @@
 
 import xpublish  # noqa: F401
 
-from .utils import TestMapper
+from .utils import TestMapper, create_dataset
 
 
-@pytest.fixture(scope="module")
+@pytest.fixture(scope='module')
 def airtemp_ds():
-    ds = xr.tutorial.open_dataset("air_temperature")
+    ds = xr.tutorial.open_dataset('air_temperature')
     return ds.chunk(dict(ds.dims))
 
 
 def test_zmetadata_identical(airtemp_ds):
     zarr_dict = {}
     airtemp_ds.to_zarr(zarr_dict, consolidated=True)
     mapper = TestMapper(airtemp_ds.rest.app)
-    actual = json.loads(mapper[".zmetadata"].decode())
-    expected = json.loads(zarr_dict[".zmetadata"].decode())
+    actual = json.loads(mapper['.zmetadata'].decode())
+    expected = json.loads(zarr_dict['.zmetadata'].decode())
     assert actual == expected
 
 
-def test_roundtrip(airtemp_ds):
-    mapper = TestMapper(airtemp_ds.rest.app)
+@pytest.mark.parametrize(
+    'start, end, freq, nlats, nlons, var_const, calendar, use_cftime',
+    [
+        ('2018-01-01', '2021-01-01', 'MS', 180, 360, True, 'standard', False),
+        ('2018-01-01', '2021-01-01', 'D', 180, 360, False, 'noleap', True),
+        ('2018-01-01', '2021-01-01', '6H', 180, 360, True, 'gregorian', False),
+        ('2018-01-01', '2050-01-01', 'A', 180, 360, None, '360_day', True),
+    ],
+)
+def test_roundtrip(start, end, freq, nlats, nlons, var_const, calendar, use_cftime):
+    ds = create_dataset(
+        start=start,
+        end=end,
+        nlats=nlats,
+        nlons=nlons,
+        var_const=var_const,
+        use_cftime=use_cftime,
+        calendar=calendar,
+    )
+    ds = ds.chunk(ds.dims)
+    mapper = TestMapper(ds.rest.app)
+    actual = xr.open_zarr(mapper, consolidated=True)
+
+    xr.testing.assert_identical(actual, ds)
+
+
+@pytest.mark.xfail(reason='Custom chunking not working properly')
```

Comment:
Any ideas what's going wrong here?

Reply:
I don't know yet. I am about to investigate.

Reply:
The slicing in `get_data_chunk()` seems to be the culprit here:


```python
chunk_id = '0.0', out_shape = (10, 2)

    def get_data_chunk(da, chunk_id, out_shape):
        """ Get one chunk of data from this DataArray (da).
    
        If this is an incomplete edge chunk, pad the returned array to match out_shape.
        """
        ikeys = tuple(map(int, chunk_id.split('.')))
        try:
            chunk_data = da.blocks[ikeys]
        except:
            chunk_data = np.asarray(da)
    
        logger.debug('checking chunk output size, %s == %s' % (chunk_data.shape, out_shape))
    
        if isinstance(chunk_data, dask_array_type):
            chunk_data = chunk_data.compute()
    
        # zarr expects full edge chunks, contents out of bounds for the array are undefined
        if chunk_data.shape != tuple(out_shape):
            new_chunk = np.empty_like(chunk_data, shape=out_shape)
            write_slice = tuple([slice(0, s) for s in chunk_data.shape])
>           new_chunk[write_slice] = chunk_data
E           ValueError: could not broadcast input array from shape (36,2) into shape (10,2)
```

Reply:
Interesting failure. I'm not 100% sure what is going wrong here. Can I ask you to refactor this function a bit before reporting back?

```python
    def get_data_chunk(da, chunk_id, out_shape):
        """ Get one chunk of data from this DataArray (da).
    
        If this is an incomplete edge chunk, pad the returned array to match out_shape.
        """
        ikeys = tuple(map(int, chunk_id.split('.')))
        
         if isinstance(chunk_data, dask_array_type):
            chunk_data = da.blocks[ikeys].compute()
        else:
            if not all([i == 0 for i in ikeys]):
                raise ValueError('xpublish does not currently supporting returning sub-array chunks for non-dask arrays')
            chunk_data = np.asarray(da)
    
        logger.debug('checking chunk output size, %s == %s' % (chunk_data.shape, out_shape))
        ...
```